File: leetcode/src/course-schedule.py
from collections import deque

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        """
        [i, j] = you must take j => i
        graph [i, j] = j --> i

        see if we can find a topological ordering
        khan's algorithm

        - take set of nodes with no incoming edges
        - remove nodes and edges attached to the nodes
        - repeat until graph is empty


        We represent graph with adj list
        """
        indegrees = [0]*numCourses
        adjList = [list() for i in range(numCourses)]

        for prereq in prerequisites:
            indegrees[prereq[0]] += 1
            adjList[prereq[1]].append(prereq[0])
        
        q = deque()
        visited = 0

        for i in range(numCourses):
            if (indegrees[i] == 0):
                q.appendleft(i)
        
        while (len(q) > 0):
            node = q.pop()
            visited += 1
            for neighbour in adjList[node]:
                indegrees[neighbour] -= 1
                if(indegrees[neighbour] == 0):
                    q.appendleft(neighbour)
        
        return visited == numCourses





        


        

File: leetcode/src/course-schedule.cpp
#include <queue>
#include <vector>
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        using namespace std;
        int visited = 0;
        vector<vector<int>> adjList (numCourses, vector<int>());
        int inDegree[numCourses];
        queue<int> q;

        for (int i = 0; i < numCourses; i ++) {
            inDegree[i] = 0;
        }

        for (auto it = prerequisites.begin(); it != prerequisites.end(); it++) {
            vector<int> edge = *it;
            adjList[edge[1]].push_back(edge[0]);
            inDegree[edge[0]]++;
        }

        for (int i = 0; i < numCourses; i ++) {
            if (inDegree[i] == 0) {
                q.push(i);
            }
        }

        while (!q.empty()) {
            int node = q.front();
            q.pop();
            visited ++;

            vector<int> neighbours = adjList[node];
            for (auto it = neighbours.begin(); it != neighbours.end(); it ++) {
                int neighbour = (*it);
                inDegree[neighbour] --;
                if (inDegree[neighbour] == 0) {
                    q.push(neighbour);
                }
            }
        }

        return (visited == numCourses);
    }
};